README.md
# Persistence API

Annotation based wrapper for SQLite and MySQL.

The library is intended for desktop applications that perform insert/update/delete operations only. 

[![BSD-2 license](https://img.shields.io/badge/License-BSD--2-informational.svg)](LICENSE)
[![Maven Central](https://maven-badges.herokuapp.com/maven-central/org.panteleyev/persistence/badge.svg)](https://maven-badges.herokuapp.com/maven-central/org.panteleyev/persistence/)
[![Javadocs](http://www.javadoc.io/badge/org.panteleyev/persistence.svg)](http://www.javadoc.io/doc/org.panteleyev/persistence)

### Artifact

```
<dependency>
    <groupId>org.panteleyev</groupId>
    <artifactId>persistence</artifactId>
    <version>19.2.0</version>
</dependency>
```

## Limitations

* Each new version may introduce incompatible changes
* No queries, either mapped or raw
* Basic support for constraints
* No support for schema migration in case of changes

## Data Types

Java | SQLite | MySQL
-----|--------|------
```String```|VARCHAR|VARCHAR
```String``` with ```Column.isJson() = true```|BLOB|JSON
```BigDecimal```|VARCHAR ( ```Column.precision() + 1``` )|DECIMAL ( ```Column.precision(), Column.scale()``` )
```Date```|INTEGER|BIGINT
```LocalDate```|INTEGER|BIGINT
```byte[]```|BLOB|VARBINARY ( ```Column.length()``` )
```UUID```|VARCHAR(36)|BINARY(16) or VARCHAR(36)

```java.util.Date``` are stored as long using ```Date.getTime()```
```java.time.LocalDate``` is stored as long using ```LocalDate.toEpochDay()```

The following types can be used as primary keys:
* Integer, int
* Long, long
* String
* UUID

Database independent auto-increment is supported for integer and long keys.

## Usage Examples

### Immutable Object

```java
@Table("book")
public class Book implements Record {
    @PrimaryKey
    @Column(Field.ID)
    private final int id;
    
    @Column("title")
    private final String title;

    @RecordBuilder
    public Book(@Column(Column.ID) int id, 
                @Column("title") String title) 
    {
        this.id = id;
        this.title = title;
    }

    public int getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }
}
```

### Foreign Keys

```java
@Table("parent_table")
public class ParentTable implements Record {
    @Column(value = "data", unique = true)
    private String data;

    public String getData() {
        return data;
    }
}

@Table("child_table")
public class ChildTable implements Record {
    @Column("parent_data")
    @ForeignKey(table = ParentTable.class, column = "data",
        onDelete = ReferenceOption.RESTRICT, onUpdate = ReferenceOption.CASCADE)
    private final String parentData;

    public String getParentData() {
        return parentData;
    }
}
```


CHANGELOG.md
# Change Log

## 19.2.0 - 2019-04-07

* MySQL: Option to store UUID as VARCHAR(36)
* Additional API methods using connection as parameter
* Unique option for @Column annotation
* Delete by id updated to support generic primary key
* Single get returns Optional (INCOMPATIBLE)

## 19.1.0 - 2019-03-23

* New annotation for primary key (INCOMPATIBLE)
* Primary key of type: Integer, Long, UUID, String
* UUID fields support
* JSON flavor of String field
* Explicit database type instead of auto-detection

## 18.2.0 - 2018-06-23

* byte[] support

## 18.1.1 - 2018-05-01

* Unsafe replaced with MethodHandle and VarHandle

## 18.1.0 - 2018-04-22

* Performance optimizations
* Java 10
* New annotations (INCOMPATIBLE)

## 3.2.1 - 2017-10-19

* Java 9 module

## 3.2.0 - 2017-10-07

* Truncate table
* Drop table
* Additional methods with connection as a parameter
* Utility methods in Record interface
* LocalDate fields support

## 3.1.0 - 2017-05-31

* Batch INSERT

### Fixed

* NPE when calling DAO.setDataSource(null)

## 3.0.0 - 2017-05-28

* MySQL support
* Database driver dependencies removed from pom.xml
* Support for indexes
* Foreign keys support fixed and fully tested
* Table listeners removed
* Updated documentation

## 2.2.1 - 2017-05-17

* Method to load database table directly to map

## 2.2.0 - 2017-02-18

* Enum fields support

## 2.1.0 - 2017-02-12

* Constants for Field annotation parameters

## 2.0.0 - 2017-02-04

* Immutable records, RecordBuilder annotation
* Support for fields of type Optional
* Internal caching removed

## 1.0.0 - 2017-01-08

* SQLite support
* Major annotations
* Mutable objects

src/main/java/org/panteleyev/persistence/MySQLProxy.java
/*
 * Copyright (c) 2017, 2019, Petr Panteleyev <petr@panteleyev.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package org.panteleyev.persistence;

import org.panteleyev.persistence.annotations.Column;
import org.panteleyev.persistence.annotations.ForeignKey;
import org.panteleyev.persistence.annotations.PrimaryKey;
import java.lang.reflect.Field;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.BiFunction;

class MySQLProxy implements DAOProxy, DAOTypes {
    private static final Map<String, BiFunction<ResultSet, String, ?>> RESULT_SET_READERS = Map.ofEntries(
        Map.entry(TYPE_STRING, OBJECT_READER),
        Map.entry(TYPE_INTEGER, OBJECT_READER),
        Map.entry(TYPE_INT, INT_READER),
        Map.entry(TYPE_LONG, OBJECT_READER),
        Map.entry(TYPE_LONG_PRIM, LONG_READER),
        Map.entry(TYPE_BOOL, BOOL_READER),
        Map.entry(TYPE_BOOLEAN, OBJECT_READER),
        Map.entry(TYPE_BIG_DECIMAL, BIG_DECIMAL_READER),
        Map.entry(TYPE_DATE, DATE_READER),
        Map.entry(TYPE_LOCAL_DATE, LOCAL_DATE_READER),
        Map.entry(TYPE_BYTE_ARRAY, BYTE_ARRAY_READER),
        Map.entry(TYPE_UUID, UUID_STRING_READER)
    );

    public Map<String, BiFunction<ResultSet, String, ?>> getReaderMap() {
        return RESULT_SET_READERS;
    }

    public String getColumnString(Column column, PrimaryKey primaryKey, ForeignKey foreignKey, String typeName,
                                  List<String> constraints)
    {
        var b = new StringBuilder();

        switch (typeName) {
            case TYPE_STRING:
                if (column.isJson()) {
                    b.append("JSON");
                } else {
                    b.append("VARCHAR(")
                        .append(column.length())
                        .append(")");
                }
                break;
            case TYPE_ENUM:
                b.append("VARCHAR(")
                    .append(column.length())
                    .append(")");
                break;
            case TYPE_BOOL:
            case TYPE_BOOLEAN:
                b.append("BOOLEAN");
                break;
            case TYPE_INTEGER:
            case TYPE_INT:
                b.append("INTEGER");
                break;
            case TYPE_LONG:
            case TYPE_LONG_PRIM:
                b.append("BIGINT");
                break;
            case TYPE_DATE:
            case TYPE_LOCAL_DATE:
                b.append("BIGINT");
                break;
            case TYPE_BIG_DECIMAL:
                b.append("DECIMAL(")
                    .append(column.precision())
                    .append(",")
                    .append(column.scale())
                    .append(")");
                break;
            case TYPE_BYTE_ARRAY:
                b.append("VARBINARY(")
                    .append(column.length())
                    .append(")");
                break;
            case TYPE_UUID:
                if (column.storeUuidAsBinary()) {
                    b.append("BINARY(16)");
                } else {
                    b.append("VARCHAR(36)");
                }
                break;
            default:
                throw new IllegalStateException(BAD_FIELD_TYPE + typeName);
        }

        if (primaryKey != null) {
            b.append(" PRIMARY KEY");
        }

        if (!column.nullable()) {
            b.append(" NOT NULL");
        }

        if (column.unique()) {
            b.append(" UNIQUE");
        }

        if (foreignKey != null) {
            constraints.add(buildForeignKey(column, foreignKey));
        }

        return b.toString();
    }

    public void truncate(Connection connection, List<Class<? extends Record>> tables) {
        try (var statement = connection.createStatement()) {
            for (var t : tables) {
                statement.execute("TRUNCATE TABLE " + Record.getTableName(t));
            }
        } catch (SQLException ex) {
            throw new RuntimeException(ex);
        }
    }

    @Override
    public void setFieldData(PreparedStatement st, int index, Object value, String typeName) throws SQLException {
        switch (typeName) {
            case TYPE_BYTE_ARRAY:
                if (value == null) {
                    st.setNull(index, Types.VARBINARY);
                } else {
                    st.setBytes(index, (byte[]) value);
                }
                break;

            default:
                DAOProxy.super.setFieldData(st, index, value, typeName);
                break;
        }
    }

    @Override
    public String getSelectColumnString(Field field) {
        var column = field.getAnnotation(Column.class);
        Objects.requireNonNull(column, FIELD_NOT_ANNOTATED + field.getName());

        if (field.getType().getTypeName().equals(TYPE_UUID) && column.storeUuidAsBinary()) {
            return "BIN_TO_UUID(" + column.value() + ") AS " + column.value();
        } else {
            return column.value();
        }
    }

    @Override
    public String getInsertColumnPattern(Field field) {
        var column = field.getAnnotation(Column.class);
        Objects.requireNonNull(column, FIELD_NOT_ANNOTATED + field.getName());

        if (field.getType().getTypeName().equals(TYPE_UUID) && column.storeUuidAsBinary()) {
            return "UUID_TO_BIN(?)";
        } else {
            return "?";
        }
    }

    @Override
    public String getUpdateColumnPattern(Field field) {
        return getInsertColumnPattern(field);
    }

    @Override
    public String getWhereColumnString(Field field) {
        var column = field.getAnnotation(Column.class);
        Objects.requireNonNull(column, FIELD_NOT_ANNOTATED + field.getName());

        if (field.getType().getTypeName().equals(TYPE_UUID) && column.storeUuidAsBinary()) {
            return "BIN_TO_UUID(" + column.value() + ")";
        } else {
            return column.value();
        }
    }
}
